### PR TITLE
feat(tools) Added a method to shift a vector with associated uncertainty

### PR DESCRIPTION
**Pull request by eichstaedtPTB**
This method allows to apply a shift (such as a time delay) to a vector and its associated uncertainty.

### CODE REVIEW

File: PyDynamic/misc/tools.py [hunk 1]
```diff
@@ -31,8 +31,45 @@
...
+
+    if isinstance(ux, float):       # no shift necessary for ux
+        return xs, ux
+    if isinstance(ux, np.ndarray):
+        if len(ux.shape) == 1:      # uncertainties given as vector
+            xs, np.roll(ux, shift)
+        if len(ux.shape) == 2:      # full covariance matrix
+            assert(ux.shape[0]==ux.shape[1])
+            uxs = np.roll(ux, shift, axis=0)
+            uxs = np.roll(uxs, shift, axis=1)
```
Comment (mgrub) [inline PyDynamic/misc/tools.py:67]: Starting from numpy 1.12.0, `np.roll` allows rolling over multiple dimensions. As we require numpy 1.19 even in our Python 3.6 `requirements.txt`, we could as well use:

```python
uxs = np.roll(ux, (shift, shift), axis=(0,1))
```

File: PyDynamic/misc/tools.py [hunk 2]
```diff
@@ -31,8 +31,45 @@
...
+
+    assert(isinstance(shift, int))
+    # application of shift to the vector of estimates
+    xs = np.roll(x, shift)
+
+    if isinstance(ux, float):       # no shift necessary for ux
+        return xs, ux
+    if isinstance(ux, np.ndarray):
+        if len(ux.shape) == 1:      # uncertainties given as vector
+            xs, np.roll(ux, shift)
```
Comment (mgrub) [inline PyDynamic/misc/tools.py:63]: Is there a `return` missing?

File: PyDynamic/misc/tools.py [hunk 3]
```diff
@@ -31,8 +31,45 @@
...
+        if len(ux.shape) == 2:      # full covariance matrix
```
Comment (mgrub) [inline PyDynamic/misc/tools.py:64]: `if` or `elif`? The difference in this case is rather subtle and should not effect the functions outcome.

**Review by mgrub — COMMENTED** [review]
Once published, we could make use of this feature in the FIRuncFilter-function to deal with the user-defined shift.

File: PyDynamic/misc/tools.py (hunk 1 above)
Comment (eichstaedtPTB) [reply PyDynamic/misc/tools.py:67]: Thanks for the great tip! I will change the code as suggested.

File: PyDynamic/misc/tools.py (hunk 2 above)
Comment (eichstaedtPTB) [reply PyDynamic/misc/tools.py:63]: Yes, indeed!
```suggestion
            return xs, np.roll(ux, shift)
```

File: PyDynamic/misc/tools.py (hunk 3 above)
Comment (eichstaedtPTB) [reply PyDynamic/misc/tools.py:64]: Using `elif` would be more appropriate.
```suggestion
        elif len(ux.shape) == 2:      # full covariance matrix
```

File: PyDynamic/misc/tools.py [hunk 4]
```diff
@@ -31,8 +31,45 @@
...
+            vector of estimates
+        ux: float, np.ndarray of shape (N,) or of shape (N,N)
+            uncertainty associated with the vector of estimates
+        shift: int
+            amount of shift
+
+    Returns
+    -------
+        np.ndarray of shape (N,)
+            shifted x and shifted ux of same type as original
```
Comment (mgrub) [inline PyDynamic/misc/tools.py:52]:
```suggestion
    Returns
    -------
        xs: (N,) array
            shifted vector of estimates
        uxs: float, np.ndarray of shape (N,) or of shape (N,N)
            uncertainty associated with the shifted vector of estimates
```

File: PyDynamic/misc/tools.py [hunk 5]
```diff
@@ -31,8 +31,45 @@
...
+
+    if isinstance(ux, float):       # no shift necessary for ux
+        return xs, ux
+    if isinstance(ux, np.ndarray):
+        if len(ux.shape) == 1:      # uncertainties given as vector
+            return xs, np.roll(ux, shift)
+        elif len(ux.shape) == 2:      # full covariance matrix
+            assert(ux.shape[0]==ux.shape[1])
+            uxs = np.roll(ux, shift, axis=0)
+            uxs = np.roll(uxs, shift, axis=1)
```
Comment (mgrub) [inline PyDynamic/misc/tools.py:67]:
```suggestion
            uxs = np.roll(ux, (shift, shift), axis=(0,1))
```

File: PyDynamic/misc/tools.py [hunk 6]
```diff
@@ -31,8 +31,45 @@
...
+            return xs, uxs
+        else:
+            raise TypeError("Input uncertainty has incompatible type")
```
Comment (mgrub) [inline PyDynamic/misc/tools.py:70]:
```suggestion
            raise TypeError("Input uncertainty has incompatible shape")
```

**Review by mgrub — COMMENTED** [review]
Some minor changes we have talked about.

File: PyDynamic/misc/tools.py [hunk 7]
```diff
@@ -31,8 +31,45 @@
     "make_equidistant",
     "trimOrPad",
     "progress_bar",
+    "shift_uncertainty"
 ]
 
+def shift_uncertainty(x, ux, shift):
+    """Shift the elements in the vector x (and associated uncertainty ux) by shift elements
+
```
Comment (eichstaedtPTB) [inline PyDynamic/misc/tools.py:39]: Amend docstring with a remark that np.roll is used for shifting.
```suggestion
    """Shift the elements in the vector x (and associated uncertainty ux) by shift elements.
        This method uses :class:`numpy.roll` to shift the elements in x and ux. See documentation of np.roll for details.

```

**Review by eichstaedtPTB — COMMENTED** [review]
Adjust docstring

**Pull request merged**